packages/liveblocks-client/src/AbstractCrdt.ts

```diff
@@ -130,7 +129,12 @@ export abstract class AbstractCrdt {
   /**
    * @internal
    */
-  abstract _serialize(parentId: string, parentKey: string, doc?: Doc): Op[];
+  abstract _serialize(
+    parentId: string,
+    parentKey: string,
+    doc?: Doc,
+    intent?: "set"
```

Comment:
That's a great question :) 

As discussed yesterday, storage operations would be easier to reason about if they were always associated with a specific LiveStructure type. 

For example, `CreateObjectOp` is sent when calling the following methods:
- `liveObject.set("key", new LiveObject());`
- `liveList.insert(0, new LiveObject());`
- `liveList.set(0, new LiveObject());`
- `liveMap.set("key", new LiveObject());`

We could have different operations for each of these methods. `LiveObjectSetOp`, `LiveListInsertOp`, etc. It's more intuitive at first glance. However, the payload of all these operations becomes more "generic"; `LiveListInsertOp` could insert a LiveObject, a LiveMap, or any other data structure that we might support in the future. This also introduces complexity for nested data structures (e.g. `liveList.insert(0, new LiveList([new LiveObject({ a : 0 })]));`) that does not exist with the existing operation structure. 

The more I think about it, the more I think we should go in that direction despite the issues above. Especially after introducing the "set intent". But since these operations are internal messages, I would wait to have more operations and data structures (LiveText, LiveBlob, etc) to refactor to a more unified solution. I just want more examples to be sure that this refactoring is really the way to go. So I share your feeling :)

Reply:
OK, perfect! Then I think we're on the same page here. We can use this API for now, and refactor this later when we have more data points and a better understanding of how to structure this 👍 Thanks for the context.